source/eventHandler.py

```diff
@@ -152,11 +152,62 @@ def _trackFocusObject(eventName, obj) -> None:
 
 	if eventName == "gainFocus":
 		lastQueuedFocusObject = obj
-		setattr(obj, WAS_GAIN_FOCUS_OBJ_ATTR_NAME, True)
-		if speech.manager._shouldDoSpeechManagerLogging():
-			log.debug(f"Changing last queued focus object: {obj!r}")
-	elif not hasattr(obj, WAS_GAIN_FOCUS_OBJ_ATTR_NAME):
-		setattr(obj, WAS_GAIN_FOCUS_OBJ_ATTR_NAME, False)
+
+
+class FocusLossCancellableSpeechCommand(_CancellableSpeechCommand):
+	def __init__(self, obj, reportDevInfo: bool):
+		from NVDAObjects import NVDAObject
+		if not isinstance(obj, NVDAObject):
+			log.warning("Unhandled object type. Expected all objects to be descendant from NVDAObject")
+			raise TypeError(f"Unhandled object type: {obj!r}")
+		self._obj = obj
+		super(FocusLossCancellableSpeechCommand, self).__init__(reportDevInfo=reportDevInfo)
+
+		if self.isLastFocusObj():
+			# Objects may be re-used.
+			# WAS_GAIN_FOCUS_OBJ_ATTR_NAME state should be cleared at some point?
+			# perhaps instead keep a weak ref list of obj that had focus, clear on keypress?
+
+			# Assumption: we only process one focus event at a time, so even if several focus events are queued,
+			# all focused objects will still gain this tracking attribute. Otherwise, this may need to be set via
+			# api.setFocusObject when globalVars.focusObject is set.
+			setattr(obj, WAS_GAIN_FOCUS_OBJ_ATTR_NAME, True)
+		elif not hasattr(obj, WAS_GAIN_FOCUS_OBJ_ATTR_NAME):
+			setattr(obj, WAS_GAIN_FOCUS_OBJ_ATTR_NAME, False)
+
+	def _checkIfValid(self):
+		stillValid = (
+			self.isLastFocusObj()
+			or not self.previouslyHadFocus()
+			or self.isAncestorOfCurrentFocus()
+			# Ensure titles for dialogs gaining focus are reported, EG NVDA Find dialog
+			or self.isForegroundObject()

```

Comment:
Is this still necessary after the change from lastQueuedFocusObject to api.getFocusObject()? I wouldn't have thought so, unless the bug is different to what we hypothesized.

Reply:
Your are right, checking for ` self.isForegroundObject()` is not required to fix this particular bug, and I have confirmed that. However, it doesn't harm the core use case of cancellable speech (rapid focus changes in Gmail), and I can't think of a reason why we would want to cancel speech for the foreground object if it wasn't the focus ancestor for some reason.

What do you think?

Reply:
I think it is redundant and should be removed, as getforegroundObject always returns a focus ancestor, which is checked already.

Reply:
I'm not sure I follow this. First, I want to confirm you want the `self.isForgroundObject()` check removed?

This may not be set by the focus ancestor, usually it is from `api.getDesktopObject().objectInForeground()`, see the following code from `doPreGainFocus`
``` python
		newForeground=api.getDesktopObject().objectInForeground()
		if not newForeground:
			log.debugWarning("Can not get real foreground, resorting to focus ancestors")
			ancestors=api.getFocusAncestors()
			if len(ancestors)>1:
				newForeground=ancestors[1]
			else:
				newForeground=obj
		api.setForegroundObject(newForeground)
```